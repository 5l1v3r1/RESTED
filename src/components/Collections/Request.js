import React from 'react';
import { connect } from 'react-redux';
import { DragSource, DropTarget } from 'react-dnd';
import { ListGroup } from 'react-bootstrap';
import flow from 'lodash.flow';

import IconButton from 'components/IconButton';
import * as Actions from 'store/collections/actions';

import { StyledRequest, AsideButtons, MainContent } from './StyledComponents';
import * as Type from './dropTypes';

/**
 * Specifies the drag source contract.
 * Only `beginDrag` function is required.
 */
const requestSource = {
  canDrag() {
    // TODO Disallow drag if editing name
    return true;
  },

  beginDrag({ id, index, collectionIndex }) {
    return { id, index, collectionIndex };
  },
};

const requestTarget = {
  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const dragCollectionIndex = monitor.getItem().collectionIndex;
    const dragUUID = monitor.getItem().id;
    const hoverIndex = props.index;
    const hoverCollectionIndex = props.collectionIndex;
    const hoverUUID = props.id;

    // Abort further processing if dragged item is hovering over itself
    if (dragUUID === hoverUUID) {
      return;
    }

    // Note: we're mutating the monitor item here.
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    /* eslint-disable no-param-reassign */
    monitor.getItem().index = hoverIndex;
    monitor.getItem().collectionIndex = hoverCollectionIndex;
    /* eslint-enable no-param-reassign */

    // Update redux orders
    props.reorderRequest({
      // Source
      collectionIndex: dragCollectionIndex,
      requestIndex: dragIndex,
    }, {
      // Target
      collectionIndex: hoverCollectionIndex,
      requestIndex: hoverIndex,
    });
  },
};

function Request(props) {
  const {
    connectDragSource,
    connectDropTarget,
    isDragging,
    id,
    collectionIndex,
    method,
    url,
    deleteRequest,
  } = props;

  return connectDragSource(connectDropTarget(
    <div> {/* Need a wrapper div for React DnD support */}
      <StyledRequest isDragging={isDragging}>
        <ListGroup componentClass="div">
          <div className="list-group-item">
            <AsideButtons>
              <IconButton
                tooltip="Edit"
                icon="cog"
              />
              <IconButton
                tooltip="Delete"
                icon="trash"
                onClick={() => deleteRequest(id, collectionIndex)}
              />
            </AsideButtons>

            <MainContent>
              <h4>{method}</h4>
              {url}
            </MainContent>
          </div>
        </ListGroup>
      </StyledRequest>
    </div>,
  ));
}

export default flow(
  DropTarget(Type.Request, requestTarget, connector => ({
    connectDropTarget: connector.dropTarget(),
  })),
  DragSource(Type.Request, requestSource, (connector, monitor) => ({
    connectDragSource: connector.dragSource(),
    isDragging: monitor.isDragging(),
  })),
  connect(null, Actions),
)(Request);

