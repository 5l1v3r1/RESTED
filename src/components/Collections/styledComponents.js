import styled, { css } from 'styled-components';
import { Panel } from 'react-bootstrap';

const hoverStyle = css`
  background-color: gray;
  color: gray;
`;

export const StyledPanel = styled.default(Panel)`
  padding-left: 0;
  padding-right: 0;
  padding-bottom: 0;

  ${props => props.isDragging && hoverStyle}

  .panel-default {
    background-color: red;
  }
`;

export const StyledRequest = styled.default.div`
  ul, li {
    ${props => props.isDragging && hoverStyle}
  }
`;

