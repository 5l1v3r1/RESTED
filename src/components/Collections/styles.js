import { StyleSheet } from 'aphrodite';

export default StyleSheet.create({
  dragPlaceholder: {
    backgroundColor: 'gray',
    color: 'gray',
  },
});

