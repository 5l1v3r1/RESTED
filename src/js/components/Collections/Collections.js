import React from 'react';
import { Col } from 'react-bootstrap';

export default function Collections() {
  return (
    <Col xsHidden sm={4}>
      <h2>
        Collections
      </h2>
    </Col>
  );
}

