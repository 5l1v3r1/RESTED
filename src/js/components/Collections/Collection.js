import React from 'react';
import { Panel } from 'react-bootstrap';
import flow from 'lodash.flow';

import Request from './Request';
import * as Type from './dropTypes';

function Collection() {
  return (
    <ul class="panel-body collection-body"
      data-ng-style="::{'display': collection.minimized ? 'none' : 'inherit'}"
      data-dnd-drop="handleDrop(item)"
      data-dnd-allowed-types="['request']"
      data-dnd-list="collection.requests">
      <Request index={0} name="foo"/>
      <Request index={1} name="bar"/>
      <Request index={2} name="baz"/>
      <Request index={3} name="qrux"/>
      <Request index={4} name="qwarp"/>
    </ul>
  );
}

export default Collection;

