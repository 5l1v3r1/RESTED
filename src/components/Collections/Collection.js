import React, { PropTypes } from 'react';
import { findDOMNode } from 'react-dom';
import { Panel } from 'react-bootstrap';
import { DragSource, DropTarget } from 'react-dnd';
import { css } from 'aphrodite';
import flow from 'lodash.flow';

import Request from './Request';
import styles from './styles';
import * as Type from './dropTypes';

import requestPropType from '../../propTypes/request';

/**
 * Specifies the drag source contract.
 * Only `beginDrag` function is required.
 */
const collectionSource = {
  beginDrag({ collectionIndex }) {
    return { collectionIndex };
  },
};

const collectionTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().collectionIndex;
    const hoverIndex = props.collectionIndex;

    // Abort further processing if dragged item is hovering over itself
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    // Note: we're mutating the monitor item here.
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    /* eslint-disable no-param-reassign */
    monitor.getItem().collectionIndex = hoverIndex;
    /* eslint-enable no-param-reassign */

    // Update redux orders
    props.reorderCollection(dragIndex, hoverIndex);
  },
};

class Collection extends React.Component {
  render() {
    const {
      requests,
      collectionIndex,
      connectDragSource,
      connectDropTarget,
      isDragging,
    } = this.props;

    return connectDragSource(connectDropTarget(
      <div>
        <Panel
          header={<h1>TODO</h1>}
          className={css(
            isDragging && styles.dragPlaceholder,
          )}
        >
          {requests.map((request, index) => (
            <Request
              key={request.id}
              index={index}
              collectionIndex={collectionIndex}
              {...request}
            />
          ))}
        </Panel>
      </div>
    ));
  }
}

Collection.propTypes = {
  requests: PropTypes.arrayOf(requestPropType).isRequired,
  collectionIndex: PropTypes.number.isRequired,
  connectDragSource: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  isDragging: PropTypes.bool.isRequired,
};


export default flow(
  DropTarget(Type.Collection, collectionTarget, connect => ({
    connectDropTarget: connect.dropTarget(),
  })),
  DragSource(Type.Collection, collectionSource, (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  })),
  /*
   * There is a weird bug where if we connect(null, Actions) here,
   * dragging between collection produces artifact requests in the
   * old collection, even though the redux store is updated. So
   * instead we pass reorderCollection as a prop, working around
   * the issue.
   *
   * It may be an issue with a redux optimization gone wrong, idk.
   */
)(Collection);

