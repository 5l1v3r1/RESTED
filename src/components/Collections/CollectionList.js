import React, { PropTypes } from 'react';
import { connect } from 'react-redux';

import Collection from './Collection';

import * as Actions from '../../store/collections/actions';
import collectionPropType from '../../propTypes/collection';

function CollectionList(props) {
  return (
    <span>
      {props.collections.map((collection, index) => (
        <Collection
          key={collection.id}
          collectionIndex={index}
          reorderCollection={props.reorderCollection}
          {...collection}
        />
      ))}
      {!props.collections.length && (
        <div>
          <h5>
            No collected requests.
            Add by pressing &quot;plus&quot; in the top right of the request panel.
          </h5>
        </div>
      )}
    </span>
  );
}

CollectionList.propTypes = {
  reorderCollection: PropTypes.func.isRequired,
  collections: PropTypes.arrayOf(collectionPropType).isRequired,
};

const mapStateToProps = ({ collections }) => ({
  collections: collections.collections,
});

export default connect(mapStateToProps, Actions)(CollectionList);

