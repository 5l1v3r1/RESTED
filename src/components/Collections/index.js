import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Panel } from 'react-bootstrap';

import * as Actions from 'store/collections/actions';
import CollectionList from './CollectionList';

function Titlebar({ addCollection }) {
  return (
    <span className="clearfix pull-right">
      <button onClick={addCollection}>
        Add new collection
      </button>
    </span>
  );
}

Titlebar.propTypes = {
  addCollection: PropTypes.func.isRequired,
};

function Collections({ addCollection }) {
  return (
    <Panel header={<Titlebar addCollection={addCollection} />}>
      <CollectionList />
    </Panel>
  );
}

Collections.propTypes = {
  addCollection: PropTypes.func.isRequired,
};

export default connect(null, Actions)(Collections);

