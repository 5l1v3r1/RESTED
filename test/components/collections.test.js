import React from 'react';
import renderer from 'react-test-renderer';

/* eslint-disable import/no-unresolved */
import Collections from 'components/Collections';

jest.mock('react-dom');

test('Collections renders correctly', () => {
  const tree = renderer.create(
    <Collections />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
