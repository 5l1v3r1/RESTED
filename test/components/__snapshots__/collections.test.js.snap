exports[`Collection component should match the previous snapshot 1`] = `
<div>
   
  <div
    className="dLMdGK">
    <span>
      <h3 />
      <button
        onClick={[Function]}>
        Delete
      </button>
      <hr />
    </span>
    <div>
       
      <div
        className="DFkvH">
        <ul
          className="list-group">
          <li
            className="list-group-item">
            <span>
              <h4>
                POST
              </h4>
              <button
                onClick={[Function]}>
                Delete
              </button>
            </span>
            <p
              className="list-group-item-text">
              www.foo.no
            </p>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collections component should match the previous snapshot 1`] = `
<div
  className="panel panel-default"
  id={undefined}>
  <div
    className="panel-heading">
    <span
      className="clearfix pull-right">
      <button
        onClick={[Function]}>
        Add new collection
      </button>
    </span>
  </div>
  <div
    className="panel-body">
    <span>
      <div>
         
        <div
          className="dLMdGK">
          <span>
            <h3>
              Collection
            </h3>
            <button
              onClick={[Function]}>
              Delete
            </button>
            <hr />
          </span>
          <div>
             
            <div
              className="DFkvH">
              <ul
                className="list-group">
                <li
                  className="list-group-item">
                  <span>
                    <h4>
                      POST
                    </h4>
                    <button
                      onClick={[Function]}>
                      Delete
                    </button>
                  </span>
                  <p
                    className="list-group-item-text">
                    www.foo.no
                  </p>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div>
         
        <div
          className="dLMdGK">
          <span>
            <h3>
              Collection 2
            </h3>
            <button
              onClick={[Function]}>
              Delete
            </button>
            <hr />
          </span>
          <div>
             
            <div
              className="DFkvH">
              <ul
                className="list-group">
                <li
                  className="list-group-item">
                  <span>
                    <h4>
                      POST
                    </h4>
                    <button
                      onClick={[Function]}>
                      Delete
                    </button>
                  </span>
                  <p
                    className="list-group-item-text">
                    www.foo.no
                  </p>
                </li>
              </ul>
            </div>
          </div>
          <div>
             
            <div
              className="DFkvH">
              <ul
                className="list-group">
                <li
                  className="list-group-item">
                  <span>
                    <h4>
                      POST
                    </h4>
                    <button
                      onClick={[Function]}>
                      Delete
                    </button>
                  </span>
                  <p
                    className="list-group-item-text">
                    www.bar.no
                  </p>
                </li>
              </ul>
            </div>
          </div>
          <div>
             
            <div
              className="DFkvH">
              <ul
                className="list-group">
                <li
                  className="list-group-item">
                  <span>
                    <h4>
                      POST
                    </h4>
                    <button
                      onClick={[Function]}>
                      Delete
                    </button>
                  </span>
                  <p
                    className="list-group-item-text">
                    www.baz.no
                  </p>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
</div>
`;
